### ISSUE TITLE
Code gets stuck when using mirrored strategy

### ISSUE THREAD

**Issue by StarDust1980** [tensorflow/tensorflow#39251]
**System information:**

Have I written custom code : Yes
OS Platform and Distribution : Linux Ubuntu 18.04
TensorFlow installed from : binary (install from pip)
TensorFlow version (use command below): 2.1.0
Python version: 3.6.8
CUDA/cuDNN version: Cuda Toolkit 10.1 / cuDNN 7.6.4
GPU model and memory: 2 Tesla V 100 16 GB each

**Question:**

I would like to store different variables and run different tensorflow ops on different devices (GPU/CPU). I want to do this because my model cannot fit on a single GPU. I am trying to train embeddings. I want to place recurring words on all GPUs and non common words on different devices (as a part of preprocessing I do all this) and transfer sentences to respective GPUs. 

For the this question I am not pasting my actual code. I am writing a snippet which is close to what I want to do. When I run the code, I do not get any error and neither does it execute successfully. I think the problem is that this code gets stuck (may be deadlock on GPU). 

Below is the snippet. For the sake of simplicity say I want to do  `y = x * w1 + x * w2 + 2 * w3`, where `w1`, `w2` and `w3` are trainable variables. I place `w1` on `GPU0` and `w2` on `GPU1`. Instead of doing the `2 * w3` on a single device say I want to split it and perform on 2 devices (to see how mirrored strategy works). I perform respective computations on respective GPUs. First of all is that a correct way to do it (using mirrored scope to assign devices to variables). Secondly if it is wrong why does tensorflow not throw any error, why is it getting stuck and why is it showing 100% GPU usage?

If i change all the device/scopes to a single GPU it executes (in seconds) successfully but I don't want that. I want to place some variables on single devices and some on both and perform computations as I illustrated below to speed up processing.
```
import os

#os.environ['TF2_BEHAVIOR'] = '1'
import tensorflow as tf
tf.compat.v1.enable_eager_execution()

#tf.compat.v1.disable_eager_execution()

tf.config.set_soft_device_placement(False)
tf.debugging.set_log_device_placement(True)
import numpy as np

class ProdLayer(tf.keras.layers.Layer):
    def __init__(self, name):
        super(ProdLayer, self).__init__()
        self.w = tf.keras.backend.variable(0.01, name='var_'+ name)


    def call(self, x):
        return x * self.w 

class SumLayer(tf.keras.layers.Layer):
    def __init__(self):
        super(SumLayer, self).__init__()
        
    def call(self, x1, x2):
        return x1 + x2
    

mirrored_strategy = tf.distribute.MirroredStrategy(devices=["/gpu:0", "/gpu:1"])

class EmbeddingModel(tf.keras.Model):
    def __init__(self):
        super(EmbeddingModel, self).__init__()
        
        # place w1 on GPU 0 and create the layer
        with tf.device('/gpu:0'):
            self.L1 = ProdLayer('w1')
        
        # place w2 on GPU 0 and create the layer        
        with tf.device('/gpu:1'):
            self.L2 = ProdLayer('w2')
        
        # place w3 on both GPU0 and GPU1 using mirrored scope. Can I do this?
        with mirrored_strategy.scope():
            self.w3 = tf.keras.backend.variable(0.01, name='var_w3')
            
        # may be do this on CPU? But for now let it perform this on GPU0
        with tf.device('/gpu:0'):
            self.L3 = SumLayer()
        
    
    def call(self, input_layer):
        # w1 is on GPU0, w2 is on GPU1 and w3 is placed using mirrored scope on both GPUs
        # y1 = w1 * x + w3
        with tf.device('/gpu:0'):
            y1 = self.L1(input_layer) + self.w3

        # y2 = w2 * x + w3
        with tf.device('/gpu:1'):
            y2 = self.L2(input_layer) + self.w3

        # y = y1 + y2 (i.e. w1 * x + w2 * x + 2 * w3) 
        with tf.device('/gpu:0'):
            y_hat = self.L3(y1, y2)
            
        return y_hat
    
    
def myLoss(y, y_hat):
    return tf.reduce_sum((y_hat-y) * (y_hat-y))

with mirrored_strategy.scope():
    model = EmbeddingModel()
    model.compile(tf.optimizers.Adam(lr=0.0001), 
                      loss=myLoss)

train_dataset = np.random.choice(100, size=(1000,))
model.fit(x=train_dataset.astype(np.float32), y=train_dataset.astype(np.float32), 
       batch_size=1000, 
       epochs=10, shuffle=False, verbose=True)

```

**Comment by ravikyram** [comment]
@StarDust1980 

Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version.Thanks!

**Comment by StarDust1980** [comment]
System information

Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Linux Ubuntu 18.04
TensorFlow installed from (source or
binary): binary (install from pip)
TensorFlow version (use command below): 2.1.0
Python version: 3.6.8
CUDA/cuDNN version: Cuda Toolkit 10.1 / cuDNN 7.6.4
GPU model and memory: 2 Tesla V 100 16 GB each

**Comment by StarDust1980** [comment]
were you able to replicate this issue on your side?

**Comment by nikitamaia** [comment]
In MirroredStrategy, your model is replicated across each GPU and the variables are kept in sync by applying identical updates. Because copies of all variables in the model are stored on each GPU, it does not sound like MirroredStrategy is suited to your use case where you want to put w1 and w2 on different GPUs.

If your main concern is that the model does not fit on your GPU, you could consider trying the [CentralStorageStrategy](https://www.tensorflow.org/api_docs/python/tf/distribute/experimental/CentralStorageStrategy), which only distributes the compute operations. This strategy will not allow you to split the variables across devices as you have suggested, but it could solve your problem of the embedding being too large to fit on GPU. 

Alternatively you could take a look at the [mesh-tensorflow library](https://github.com/tensorflow/mesh) which is designed for model parallelism, where you are splitting the model itself between different devices. If you are interested in learning more about this library, there is a [fifteen minute introduction](https://www.youtube.com/watch?v=HgGyWS40g-g) from the TF Dev Summit in 2019.

**Comment by anj-s** [comment]
@StarDust1980 Have you tried using a [custom training loop(CTL)](https://www.tensorflow.org/tutorials/distribute/custom_training) instead of the native tf.keras `compile` and `fit`? Since you are explicitly placing variables and calling it on specific devices I would suggest using a CTL to give you more extensibility.
If you place any variable creation under MirroredStrategy scope then it is replicated on the devices that you used to instantiate the strategy with. I don't know how `compile` will behave since we haven't tested this setup which is why you may be seeing your training hang in this setup.

**Comment by StarDust1980** [comment]
I was mainly trying to figure out if I can have some variables (like w1 and w2) on independent GPUs and some replicated (like w3 - using mirrored strategy). I remember trying CTL but even that froze. Can this be done at all? I mean using mirrored strategy for some trainable variables. In the documentation, you copy all the variables on all the GPUs. I do not want to do that. I just need parts to be replicated.

If it can be done using mirrored strategy, can you provide a dummy snippet which does that (w1 on GPU1, w2 on GPU2 and w3 on both GPU1 and GPU2 using mirroring). The loss/use case doesn't have to make sense. Just want to know if that can be done and how?

**Comment by anj-s** [comment]
@StarDust1980 Here is a snippet that works for me. Notice the commented out code:
`
import tensorflow as tf
gpus = tf.config.experimental.list_physical_devices('GPU')
tf.config.experimental.set_virtual_device_configuration(gpus[0], [
  tf.config.experimental.VirtualDeviceConfiguration(memory_limit=1024),
  tf.config.experimental.VirtualDeviceConfiguration(memory_limit=1024),
])
mirrored_strategy = tf.distribute.MirroredStrategy(devices=["/gpu:0", "/gpu:1"])

class EmbeddingModel(tf.keras.Model):
    def __init__(self):
        super(EmbeddingModel, self).__init__()
        
        # place w1 on GPU 0 and create the layer
        with tf.device('/gpu:0'):
            self.L1 = ProdLayer('w1')
        
        # place w2 on GPU 0 and create the layer        
        with tf.device('/gpu:1'):
            self.L2 = ProdLayer('w2')
        
        # place w3 on both GPU0 and GPU1 using mirrored scope. Can I do this?
        with mirrored_strategy.scope():
            self.w3 = tf.keras.backend.variable(0.01, name='var_w3')
            
        # may be do this on CPU? But for now let it perform this on GPU0
        with tf.device('/gpu:0'):
            self.L3 = SumLayer()
        
    
    def call(self, input_layer):
        # w1 is on GPU0, w2 is on GPU1 and w3 is placed using mirrored scope on both GPUs
        # y1 = w1 * x + w3
        with tf.device('/gpu:0'):
            y1 = self.L1(input_layer) + self.w3

        # y2 = w2 * x + w3
        with tf.device('/gpu:1'):
            y2 = self.L2(input_layer) + self.w3

        # y = y1 + y2 (i.e. w1 * x + w2 * x + 2 * w3) 
        with tf.device('/gpu:0'):
            y_hat = self.L3(y1, y2)
            
        return y_hat
    
    
def myLoss(y, y_hat):
    return tf.reduce_sum((y_hat-y) * (y_hat-y))

# with mirrored_strategy.scope(): # This cannot work
    # model = EmbeddingModel()
    # model.compile(tf.optimizers.Adam(lr=0.0001), 
    #                   loss=myLoss)

train_dataset = np.random.random((10))
model = EmbeddingModel()
model(train_dataset)
# Output
# <tf.Tensor: shape=(10,), dtype=float64, numpy=
# array([0.02833105, 0.03860711, 0.02790513, 0.02937286, 0.03635627,
#       0.02511177, 0.03697144, 0.03635903, 0.03183693, 0.03885038])>
`

You can't place EmbeddingModel in strategy scope after specifying the variable scope explicitly. This might have been why the training was not making any progress. Hope this helps unblock your use case.

**Comment by nikitamaia** [comment]
Hope this was helpful, @StarDust1980. Closing this issue now since an example has been provided.
